lib/mongodb.js

```diff
@@ -304,29 +304,40 @@ MongoDB.prototype.connect = function(callback) {
       err,
       client
     ) {
-      if (!err) {
+      if (err) {
+        if (callback) callback(err);
         if (self.debug) {
-          debug('MongoDB connection is established: ' + self.settings.url);
-        }
-        self.client = client;
-        // The database name might be in the url
-        return urlParser(self.settings.url, self.settings, function(err, url) {
-          self.db = client.db(
-            url.dbName || self.settings.database,
-            url.db_options || self.settings
-          );
-          if (callback) callback(err, self.db);
-        });
-      } else {
-        if (self.debug || !callback) {
           g.error(
             '{{MongoDB}} connection is failed: %s %s',
             self.settings.url,
             err
           );
         }
```

Comment:
Have you considered extracting this long block of error-handling code into a shared function inside `connect`? 

```js
if (err) return onError(err);
self.client = client;
// The database name might be in the url
return urlParser(self.settings.url, self.settings, function(err, url) {
  if (err) onError(err);
  // success - carry on
}

function onError(err) {
  if (callback) callback(err);
  if (self.debug) {
    // ...
  }
}
```

Also is it a good idea to pass the error to callback before logging it? Wouldn't it produce confusing logs, where the caller prints its error information and he low-level details are printed only later? Personally, I would log the error first and call back second.